lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+            address: The address of the vault server.
+
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the vault url without being the leader. Ignoring the request."
+            )
+            return
+        relation.data[self.charm.app].update({"address": address})
+
+    def set_credentials_secret_id(self, relation: ops.Relation, secret_id: str) -> None:
+        """Set the credentials secret id in the relation databag.
+
+        Args:
+            relation: The relation for which to set the credentials secret id.
+            secret_id: The secret id of the Juju secret which stores the credentials for authenticating with the Vault server.
+        """
+        if relation.app is None:
+            raise ValueError("The `relation.app` is not set")
+        relation.data[self.charm.app].update({"credentials_secret_id": secret_id})
+
+    def set_ca_certificate(self, relation: ops.Relation, ca_certificate: str) -> None:
+        """Set the ca_certificate on the relation.
+
+        Args:
+            relation: The relation for which to set the ca_certificate.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if not relation:
+            logger.warning("Relation is None")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        relation.data[self.charm.app]["ca_certificate"] = ca_certificate
+
+
+def is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Return whether the provider data is valid.
+
+    This uses the pydantic schema to validate the data.
+
+    Args:
+        data: The data to validate.
+
+    Returns:
+        True if the data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        if event.app is None:
+            logger.warning("No remote application yet")
+            return
+        if is_provider_data_valid(event.relation.data[event.app]):
+            self.on.vault_autounseal_details_ready.emit(event.relation)
+
+    def get_vault_url(self, relation: ops.Relation) -> Optional[str]:
```

Comment:
Why not take fetch the relation directly instead of asking it as a parameter? We already know the relation name from the `__init__`. This applies at a couple of places.

Reply:
We could. I assumed there would be some performance benefit to not retrieving this every time. Once again following precedence in vault_kv.

Reply:
I like passing it in as a parameter because we don't have to repeat `if not` statements and rely on the static type checker to warn if it's `type | None` if used improperly.

Reply:
fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/0651467cd3a75065ba0f04391194ceef96844526